### ISSUE TITLE
Anonymous users can see schema structure they shouldn't

### ISSUE THREAD

**Issue by begriffs** [begriffs/postgrest#453]
Making a GET request to the root path shows tables that the anonymous user does not have access to read or write.

Steps to reproduce:

```bash
createdb perms
psql -d perms
```

then in psql
```sql
create role perm_auth noinherit login;
create role perm_anon;
create role perm_user;
create table festival
(
  name text not null primary key
);
grant perm_anon to perm_auth;
grant perm_user to perm_auth;
grant usage on schema public to perm_user;
```

Start postgrest
```bash
postgrest postgres://perm_auth:@localhost:5432/perms -a perm_anon --schema public
```

At this point when I ask for `/` it gives me back an empty array which is good. Nothing is visible to the `perm_anon` user. Then I go back into psql and execute `grant all on festival to perm_user;` (not `perm_anon`). Requesting `/` now gives me

```json
[
  {
    "schema": "public",
    "name": "festival",
    "insertable": true
  }
]
```

**Comment by ruslantalpa** [comment]
So the query that works correctly i think is here
https://github.com/begriffs/postgrest/blob/3f1d2d8ed9a574cf9b97259334a2f1b6cb98cfae/src/PostgREST/PgStructure.hs#L107

But the logic changed around this PR https://github.com/begriffs/postgrest/commit/0374d4e6513728ff7c2fa0847a977f2e13f11f31
and replaced with the query from accessibleTables. That query probably was from an intermediate commit but it's output was filtered by a more complex function that checked the roles that had access to that table.

Long story short, if the query mentioned in the beginning is added back to accessibleTables and the query does all the work (no filtering after the output) i think it will work.
and the type of accessibleTables should be ```Schema -> H.Tx P.Postgres s [Table]```

I'll try a PR on monday (in no one gets to it before than :) )

**Comment by ruslantalpa** [comment]
Thing got lost when we switched from getting the tables first from db, then from cache, then back

**Comment by ruslantalpa** [comment]
Investigating this, at first look it seems that "postgrest_test_author" role and "author_only" table is modeling exactly this situation. One question, after you do "grat all" did you restart the postgrest instance? My guess is that this error occurs only when you do the grat all thing but do not restart the server. Anyway, i will add that PR soon

**Comment by begriffs** [comment]
Good point, I didn't restart the server after changing permissions.

**Comment by begriffs** [comment]
So this is weird: I tried this again to see if the problem would persist if I restarted postgrest after changing the schema but the problem didn't happen at all, even without a restart! I'm confused it happened in the first place. I tried dropping the db and the users and starting fresh. Tried with master, 0.3.0.3, and 0.3.0.2.

@ruslantalpa can you try to reproduce as well please?
